test/test-parse-warning.cpp

```diff
@@ -342,6 +342,86 @@ TEST_F(test_parse_warning,
   }
 }
 
+TEST_F(test_parse_warning,
+       warn_on_comma_between_member_array_subscript_operators) {
+  {
+    test_parser p(u8"a[1, 2, 3]"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(p.code,
+                              diag_misleading_comma_operator_in_index_operation,
+                              comma, strlen(u8"a[1, 2"), u8","_sv),
+        }));
+  }
+  {
+    test_parser p(u8"sqrt(a[1, 2, 3])"_sv, capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(p.code,
+                              diag_misleading_comma_operator_in_index_operation,
+                              comma, strlen(u8"sqrt(a[1, 2"), u8","_sv),
+        }));
+  }
+  {
+    test_parser p(u8"a[pow(1, 2), 2]"_sv, capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(p.code,
+                              diag_misleading_comma_operator_in_index_operation,
+                              comma, strlen(u8"a[pow(1, 2)"), u8","_sv),
+        }));
+  }
+  {
+    test_parser p(u8"a[b[1967, 1975]]"_sv, capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(p.code,
+                              diag_misleading_comma_operator_in_index_operation,
+                              comma, strlen(u8"a[b[1967"), u8","_sv),
+        }));
+  }
+  {
+    test_parser p(u8"let b = a[-1, 1]"_sv, capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(p.code,
+                              diag_misleading_comma_operator_in_index_operation,
+                              comma, strlen(u8"let b = a[-1"), u8","_sv),
+        }));
+  }
+  {
+    test_parser p(u8"a[0][-1, 1]"_sv, capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(p.code,
+                              diag_misleading_comma_operator_in_index_operation,
+                              comma, strlen(u8"a[0][-1"), u8","_sv),
+        }));
+  }
+  {
+    test_parser p(u8"a = [1, 2, 3]"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(p.errors, IsEmpty());
+  }
+  {
+    test_parser p(u8"for (let i = 0, j = 1; ; ++i, ++j) {}"_sv, capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(p.errors, IsEmpty());
+  }
+}
```

Comment:
Nitpick: I think this testing is excessive for how simple the implementation is. I'd deduplicate some tests, such as the `a[b[1967, 1975]]` and `a[0][-1, 1]` and `let b = a[-1, 1]` tests which cover the same thing.